### PR TITLE
Add ChangeLogModal

### PR DESCRIPTION
**Pull request by kapillamba4**
### Description
Added Change Log modal

### Screenshots
captured

### CODE REVIEW

File: frontend/src/component/styles/color.scss [hunk 1]
```diff
@@ -1,7 +1,11 @@
 $color-primary: #e91e63;
 $color-white: #fff;
+$color-white-smoke: #f5f5f5;
```
Comment (byliuyang) [inline frontend/src/component/styles/color.scss]:
```suggestion
$color-modal-background: #f5f5f5;
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+.modal-wrapper {
+  left: 0%;
+  position: fixed;
+  top: 0%;
+  width: 100%;
+  height: 100%;
+
+  .modal-body {
+    background-color: $color-white;
+    border: 1px solid $color-light-gray-1;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    left: 50%;
+    position: fixed;
+    top: 50%;
+    width: 500px;
+    height: 500px;
+    transform: translateX(-50%) translateY(-50%);
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]: Let's create a mixin for align center CSS:

```scss
@mixin align-center {
    top: 50%;
    left: 50%;
    transform: translateX(-50%) translateY(-50%);
}
```

File: frontend/src/component/styles/color.scss [hunk 4]
```diff
@@ -1,7 +1,11 @@
...
 $color-dark-gray: #343434;
 $color-highlight: lightblue;
+$color-light-gray-0: #cccccc;
+$color-light-gray-1: #dbdbdb;
```
Comment (byliuyang) [inline frontend/src/component/styles/color.scss]:
```suggestion
$color-divider: #dbdbdb;
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    ul {
+      height: calc(500px - 114px);
+      list-style: none;
+      display: block;
+      padding: 0;
+      margin: 0;
+      overflow-y: scroll;
+      background-color: $color-white-smoke;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      border: 4px solid $color-white-smoke;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        border: 1px solid $color-light-gray-1;
+        margin: 10px;
+        min-height: 120px;
+        background: white;
+        position: relative;
+        word-break: break-all;
+
+        .title {
+          font-weight: 900;
+          padding-bottom: 6px;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
          margin-bottom: 6px;
```
Padding counts as part of the height of title. I guess you want to keep distance between title and description. Please checkout CSS box modal: https://www.w3schools.com/css/css_boxmodel.asp

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      li {
+        padding: 10px;
+        list-style: none;
+        text-align: left;
+        border: 1px solid $color-light-gray-1;
+        margin: 10px;
+        min-height: 120px;
+        background: white;
+        position: relative;
+        word-break: break-all;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]: Nit: Should we break all? Will this affect readability?

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+        .title {
+          font-weight: 900;
+          padding-bottom: 6px;
+        }
+
+        .published-date {
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
        released-date {
```

File: frontend/src/component/styles/color.scss [hunk 10]
```diff
@@ -1,7 +1,11 @@
...
+$color-light-gray-2: #999999;
```
Comment (byliuyang) [inline frontend/src/component/styles/color.scss]:
```suggestion
$color-time: #999999;
```

File: frontend/src/component/styles/color.scss [hunk 11]
```diff
@@ -1,7 +1,11 @@
 $color-primary: #e91e63;
 $color-white: #fff;
+$color-white-smoke: #f5f5f5;
 $color-dark-gray: #343434;
 $color-highlight: lightblue;
+$color-light-gray-0: #cccccc;
```
Comment (byliuyang) [inline frontend/src/component/styles/color.scss]:
```suggestion
$color-scrollbar-track: #cccccc;
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    if (changeLog) {
+      return (
+        <ul
+          className={classnames({
+            'complete-list': this.state.shouldShowCompleteChangeLog
+          })}
+        >
+          {changeLog.map((update: Update) => (
+            <li key={update.publishedAt}>
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
            <li key={update.releasedAt}>
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  showCompleteChangeLog = () => {
+    this.setState({
+      shouldShowCompleteChangeLog: true
+    });
+  };
+
+  createChangeLog = () => {
+    let changeLog = this.props.changeLog;
+    if (!this.state.shouldShowCompleteChangeLog) {
+      changeLog = changeLog!.slice(0, 3);
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
      changeLog = changeLog!.slice(0, this.props.defaultVisibleLogs);
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+          })}
+        >
+          {changeLog.map((update: Update) => (
+            <li key={update.publishedAt}>
+              <div className={'title'}>{update.title}</div>
+              <div>{update.excerpt}</div>
+              <br />
+              <div className={'published-date'}>
+                {moment(update.publishedAt).fromNow()} -{' '}
+                {moment(update.publishedAt).format('MMMM Do YYYY, h:mm:ss a')}
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
                {moment(update.publishedAt).fromNow()}
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+          className={classnames({
+            'complete-list': this.state.shouldShowCompleteChangeLog
+          })}
+        >
+          {changeLog.map((update: Update) => (
+            <li key={update.publishedAt}>
+              <div className={'title'}>{update.title}</div>
+              <div>{update.excerpt}</div>
+              <br />
+              <div className={'published-date'}>
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
              <div className={'released-date'}>
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+      return (
+        <ul
+          className={classnames({
+            'complete-list': this.state.shouldShowCompleteChangeLog
+          })}
+        >
+          {changeLog.map((update: Update) => (
+            <li key={update.publishedAt}>
+              <div className={'title'}>{update.title}</div>
+              <div>{update.excerpt}</div>
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
              <div class="summary">{update.summary}</div>
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        min-height: 120px;
+        background: white;
+        position: relative;
+        word-break: break-all;
+
+        .title {
+          font-weight: 900;
+          padding-bottom: 6px;
+        }
+
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
.summary {
    flex: 1;
}
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+      display: block;
+      padding: 0;
+      margin: 0;
+      overflow-y: scroll;
+      background-color: $color-white-smoke;
+      border: 4px solid $color-white-smoke;
+      @include scroll-bar;
+
+      li {
+        padding: 10px;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
      li {
        display: flex;
        padding: 10px;
```

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 17 above)
Comment (byliuyang) [reply frontend/src/component/ui/ChangeLogModal.scss]: This allows summary to take the rest of spaces between title and time.

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        position: relative;
+        word-break: break-all;
+
+        .title {
+          font-weight: 900;
+          padding-bottom: 6px;
+        }
+
+        .published-date {
+          position: absolute;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+          bottom: 10px;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
          margin-bottom: 10px;
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+          color: $color-light-gray-2;
+        }
+      }
+    }
+
+    ul.complete-list {
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
    ul.full-list {
```

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+      <Button onClick={this.showCompleteChangeLog}>View All Updates</Button>
+    );
+  };
+
+  render() {
+    if (!this.props.shouldShowModal) {
+      return <div />;
+    }
+    return (
+      <div className={'modal-wrapper'}>
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]: We can reuse this modal as the wrapper: https://github.com/short-d/short/blob/master/frontend/src/component/ui/Modal.tsx

File: frontend/src/component/ui/ChangeLogModal.tsx [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+    return <div />;
+  };
+
+  createShowCompleteChangeLogButton = () => {
+    if (this.state.shouldShowCompleteChangeLog) {
+      return <div />;
+    }
+    return (
+      <Button onClick={this.showCompleteChangeLog}>View All Updates</Button>
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.tsx]:
```suggestion
      <Button onClick={this.showFullChangeLog}>View All Updates</Button>
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+      border-bottom: 1px solid $color-light-gray-1;
+      font-weight: 600;
+
+      i.clear {
+        cursor: pointer;
+      }
+    }
+
+    ul {
+      height: calc(500px - 114px);
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
flex: 1;
```

File: frontend/src/component/ui/ChangeLogModal.scss [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+
+.modal-wrapper {
+  left: 0%;
+  position: fixed;
+  top: 0%;
+  width: 100%;
+  height: 100%;
+
+  .modal-body {
+    background-color: $color-white;
```
Comment (byliuyang) [inline frontend/src/component/ui/ChangeLogModal.scss]:
```suggestion
  .modal-body {
    display: flex;
    flex-direction: column;
    background-color: $color-white;
```

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 24 above)
Comment (byliuyang) [reply frontend/src/component/ui/ChangeLogModal.scss]: This and combined with:

```scss
 display: flex;
 flex-direction: column;
```
will automatically determine the size of the `ul`

**Comment by byliuyang** [comment]
Hi @kapillamba4, thank you so much for working this awesome feature! I left some comments for the first round of code review. We will go for the second round after the current ones are addressed.

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 5 above)
Comment (kapillamba4) [reply frontend/src/component/ui/ChangeLogModal.scss]: Why remove background from ul? there is background color in [this image](https://user-images.githubusercontent.com/3537801/69011841-ae3d9c00-0923-11ea-9821-24ea5e0318c6.png)

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 6 above)
Comment (kapillamba4) [reply frontend/src/component/ui/ChangeLogModal.scss]: It is to give padding to scrollbar, i cannot find another way is there to do it.

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 5 above)
Comment (byliuyang) [reply frontend/src/component/ui/ChangeLogModal.scss]: Oh, I misunderstand.

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 8 above)
Comment (kapillamba4) [reply frontend/src/component/ui/ChangeLogModal.scss]: I remember changing it to break-word, it seems i stashed the change accidentally.

File: frontend/src/component/ui/ChangeLogModal.scss (hunk 6 above)
Comment (byliuyang) [reply frontend/src/component/ui/ChangeLogModal.scss]: You can add margin to the summary section.

**Pull request merged**